adapters/mediago/mediago.go

```diff
@@ -0,0 +1,222 @@
+package mediago
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"text/template"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+type adapter struct {
+	EndpointTemplate *template.Template
+}
+
+type mediagoResponseBidExt struct {
+	MediaType string `json:"mediaType"`
+}
+
+// Builder builds a new instance of the MediaGo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	endpoint, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+	bidder := &adapter{
+		EndpointTemplate: endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+	var errs []error
+
+	adapterRequest, err := a.makeRequest(request)
+	if err == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	} else {
+		errs = append(errs, err)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	mediagoExt, err := getImpMediaGoExt(&request.Imp[0])
+
+	if err != nil {
+		return nil, &errortypes.BadInput{Message: "Invalid ExtImpMediaGo value"}
+	}
+
+	endPoint, err := a.getEndPoint(mediagoExt)
+	if err != nil {
+		return nil, err
+	}
+
+	preProcess(request)
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+// get ImpMediaGoExt From First Imp. Only check and get first Imp.Ext.Bidder to ExtImpMediago
+func getImpMediaGoExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpMediaGo, error) {
+	var extImpMediaGo openrtb_ext.ExtImpMediaGo
+	var extBidder adapters.ExtImpBidder
+	err := json.Unmarshal(imp.Ext, &extBidder)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(extBidder.Bidder, &extImpMediaGo)
+	if err != nil {
+		return nil, err
+	}
+	return &extImpMediaGo, nil
+}
+
+func getRegionInfo(region string) string {
+	switch region {
+	case "APAC":
+		return "jp"
```

Comment:
code part is not covered. should add json test with `APAC` region for better coverage